### DISCUSSION TITLE
plot Sphere with angle theta covering 0

### DISCUSSION THREAD

**Discussion by chapochn** [pyvista/pyvista#6453] [Q&A]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [x] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
import pyvista as pv

pl = pv.Plotter()
a1 = 180+90
sphere1 = pv.SolidSphere(direction=[0, 0, 1], start_theta=a1, end_theta=a1+180)
sphere2 = pv.Sphere(direction=[0, 0, 1], start_theta=a1, end_theta=a1+180, center=[1, 0, 0])
pl.add_mesh(sphere1)
pl.add_mesh(sphere2)

pl.view_xy(negative=False)
pl.show()
```

### Description

I understand that in the documentation, theta for Sphere is only allowed to be between 0 and 360. What is the suggested way to have a section that covers the angle 0 just like for SolidSphere? Could that limitation be lifted in the code?
I understand that one workaround would be to use the rotation function. Any other options?

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Thu Aug 01 17:00:35 2024 EDT

                OS : Darwin (macOS 14.5)
            CPU(s) : 12
           Machine : arm64
      Architecture : 64bit
               RAM : 32.0 GiB
       Environment : Jupyter
       File system : apfs
        GPU Vendor : Apple
      GPU Renderer : Apple M2 Pro
       GPU Version : 4.1 Metal - 88.1
  MathText Support : True

  Python 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:51:49)
  [Clang 16.0.6 ]

           pyvista : 0.44.1
               vtk : 9.3.1
             numpy : 1.26.4
        matplotlib : 3.9.1
            scooby : 0.10.0
             pooch : 1.8.2
            pillow : 10.4.0
           imageio : 2.34.2
         pyvistaqt : 0.11.1
             PyQt5 : 5.15.9
           IPython : 8.26.0
        ipywidgets : 8.1.3
             scipy : 1.14.0
              tqdm : 4.66.4
        jupyterlab : 4.2.4
             trame : 3.6.3
      trame_client : 3.2.1
      trame_server : 3.0.3
         trame_vtk : 2.8.9
     trame_vuetify : 2.6.2
      nest_asyncio : 1.6.0
--------------------------------------------------------------------------------
```

### Screenshots

<img width="690" alt="Screenshot 2024-08-01 at 5 00 57 PM" src="https://github.com/user-attachments/assets/20e5b1c0-e567-47c4-adb8-ce334cb9e147">

**Comment by user27182** [comment]
I suspect this is a limitation with how vtk generates the sphere.

A workaround is to first "rotate" the input angles to ensure the upper range does not exceed 360, and then rotate the output by the same amount.

``` python
import pyvista as pv

a1 = 270
a2 = a1 + 180

### Limit upper range of theta to 360
angle_offset = a2 % 360
if angle_offset:
    a1_new = a1 - angle_offset
    a2_new = a2 - angle_offset

solid_sphere = pv.SolidSphere(start_theta=a1, end_theta=a2)
center= [1,0,0]
surf_sphere = pv.Sphere(start_theta=a1_new, end_theta=a2_new, center=center)

# Rotate by offset applied earlier
if angle_offset:
    surf_sphere.rotate_z(angle_offset, point=center, inplace=True)

pl = pv.Plotter()
pl.add_mesh(solid_sphere)
pl.add_mesh(surf_sphere)
pl.view_xy(negative=False)
pl.show()
```
![image](https://github.com/user-attachments/assets/8c50a933-644c-491b-9e3b-d6fa28d798b9)

**Reply by chapochn** [reply]
Thanks, yes, that was one of the workarounds I wrote above and what I implemented... I was wondering if there were other options or a possibility to lift that limitation in pyvista...

**Reply by user27182** [reply]
> Thanks, yes, that was one of the workarounds I wrote above and what I implemented... I was wondering if there were other options or a possibility to lift that limitation in pyvista...

I personally don't have any other ideas for making this work.

Feel free to modify the source code for `Sphere` directly and add the workaround, and submit and PR with the changes to remove this limitation.

**Reply by MatthewFlamm** [reply]
> Feel free to modify the source code for `Sphere` directly and add the workaround, and submit and PR with the changes to remove this limitation.

One option would be to do starting point and end point detection and shifting in the `pyvista.Sphere` or `pyvista.SphereSource` code. This would be straightforward to implement.  The code in `pyvista.SolidSphere` could be used as a guide, although there are a lot handling needed there.

`pyvista.Sphere` uses `vtkSphereSource`, see [vtk docs for vtkSphereSource](https://vtk.org/doc/nightly/html/classvtkSphereSource.html#details).  `pyvista.SolidSphere` is written directly for PyVista so it has a few more bells and whistles.

I am working on this example showing how a 2D surface sphere mesh can be constructed using numpy and pyvista.    See https://github.com/pyvista/pyvista/pull/6450.  If this was made into a pyvista geometric object like `pyvista.QuadSphere` or something like this, we could have an interface be the same as `SolidSphere`.
